File: leetCode/python/jan2023/snakesAndLadders.py
from typing import List
from heapq import *
from collections import deque

class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        maxIndex = len(board) ** 2 - 1
        frontier = deque()
        frontier.append(0)
        n = len(board)
        startingRowMod = (n - 1) % 2
        currentSteps = 0
        newFrontier = deque()
        visited = set()
        while (frontier):
            currentStart = frontier.pop()
            if currentSteps > maxIndex:  return -1
            for i in range(1, 7):
                nextStep = currentStart + i
                if nextStep in visited: continue
                visited.add(nextStep)
                if nextStep > maxIndex:  break
                rowIndex = n - nextStep // n - 1
                if rowIndex % 2 == startingRowMod:
                    colIndex = nextStep % n
                else:
                    colIndex = n - (nextStep % n) - 1
                nextValue = board[rowIndex][colIndex]
                if nextValue != -1:
                    nextStep = nextValue - 1
                if nextStep == maxIndex:
                    return currentSteps + 1
                newFrontier.append(nextStep)
                
            if len(frontier) == 0:
                frontier = newFrontier
                newFrontier = deque()
                currentSteps += 1
        return -1

def main():
    sol = Solution()
    board = [[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,35,-1,-1,13,-1],[-1,-1,-1,-1,-1,-1],[-1,15,-1,-1,-1,-1]]
    result = sol.snakesAndLadders(board)
    print(result)
    assert result == 4
    board = [[-1,-1,-1],[-1,9,8],[-1,8,9]]
    result = sol.snakesAndLadders(board)
    print(result)
    assert result == 1
    board = [[-1,-1,2,-1],[14,2,12,3],[4,9,1,11],[-1,2,1,16]]
    result = sol.snakesAndLadders(board)
    print(result)
    assert result == 1
    board = [[-1,-1],[-1,3]]
    result = sol.snakesAndLadders(board)
    print(result)
    assert result == 1
    board = [[1,1,-1],[1,1,1],[-1,1,1]]
    result = sol.snakesAndLadders(board)
    print(result)
    assert result == -1

main()

'''
Cycles are a problem
Depth first is a problem
Switching to priority queue solves both?

Indexing is still off
0 always maps to colIndex = 0
So can't use hardcored modulus value based on the row
But if we know starting row modulus, then we set things

Simpler question: formula for swtiched colIndex
    Default: nextStep % n 
    reverse: 0 -> n - 1
            1 -> n - 2
            i -> n - i - 1

My version is a lot slower than most
    Worst case time complexity:
        Unreachable
        Need to go up to n^2 steps on every path
    Can improve for some cases by checking if last square is directly accessible
        Doesn't help with breaks that happen later though

Better solution ideas:
    Priority queue from end to start
    Adjacency list

Adjacency list:
    Node adjacent to:
        Any that directly reference it
        i - 1, if i - 1 == -1

Problem: for non-normal rolls, adjacency list doesn't read the intermediate but the starter
    And that starter can be more then one
    E.g. if a bunch can reach index k, which is a snake or ladder
        Then the destination it goes to is actually connected by 1 step to the other ones
            But how does a double jump work there?
            By only including the direct one, you're covered
                I.e. don't do it recursively

Have a second adjacency list for get there by snake or ladder?
'''

File: leetCode/typescript/jan2023/snakesAndLadders.ts
function snakesAndLadders(board: number[][]): number {
    var minSteps:number = board.length ** 2 - 1;
    // var frontier:number[][] = [0, board[n - 1][0]];
    minSteps = recursiveSnakesAndLadders(board, 0, false)
    return minSteps;
};

function recursiveSnakesAndLadders(board: number[][], start:number, didSpecialMove:boolean): number {
    var maxIndex:number = (board.length ** 2) - 1;
    var minSteps:number = maxIndex;
    for (var i=1; i < 7; i++) {
        var nextStep:number = start + i
        if (nextStep > maxIndex) { continue }
        if (nextStep === maxIndex) {
            console.log(`arrived from ${start}`)
            return 1
        } else {
            var newDidSpecialMove:boolean = false
            if (!didSpecialMove) {
                var rowIndex = board.length - Math.floor(nextStep / board.length) - 1
                var colIndex:number;
                console.log(rowIndex)
                if (rowIndex % 2 != 0) {
                    colIndex = nextStep % board.length
                } else {
                    colIndex = board.length - (nextStep % board.length) - 1
                }
                var destinationValue = board[rowIndex][colIndex]
                if (destinationValue !== -1 && !didSpecialMove) {
                    newDidSpecialMove = true
                    // Minus 1 to get index, not label
                    nextStep = destinationValue - 1
                }
            }
            var newMinSteps = recursiveSnakesAndLadders(board, nextStep, newDidSpecialMove) + 1
            minSteps = Math.min(minSteps, newMinSteps)
            if (start === 0)  {
                console.log(`nextStep ${nextStep}`)
                console.log(`minSteps ${minSteps}`)
            }
        }
    }
    return minSteps
}

function snakesAndLaddersMain(): void {
    var board:number[][];
    var resultNum:number;
    board = [[-1,-1],[-1,3]]
    resultNum = snakesAndLadders(board)
    console.log(resultNum)
    console.assert(resultNum === 1)
    return
    board = [[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,35,-1,-1,13,-1],[-1,-1,-1,-1,-1,-1],[-1,15,-1,-1,-1,-1]]
    resultNum = snakesAndLadders(board)
    console.log(resultNum)
    console.assert(resultNum === 4)
}

snakesAndLaddersMain()

/*
Data range/assumptions:
n x n
n: [2, 20]
Boustrophedon style
grid values: [-1] U [1, n^2]
*/

/*
Tests:
n = 2
n = 20
double snake
double ladder
snake needed in optimal path
cycles if using dfs
*/

/*
Ideas:

Naive:
    Recursive, BFS/priority queue
    Choose all moves and take the minimum

    Worst case: no snakes or ladders?

Better: work back from end using a sort of adjacency list 
    Collect any tiles that lead to the current tile
    If none, use the last tile
    Repeat
    Is it reliably better?
        Still branches a lot
        Natural priority structure, since any square one away should be prioritized over 2 away, etc.

Prioritize ladders?

Index from space:

row = n - index // current - 1
column is different if row is even or odd
    even: index % current
    odd: n - index % current

n = 2
row: 0 -> 2 - 0 - 1 -> 1
col: 0 -> row is odd -> 0 % 2 -> 0

1
1 // 2 -> 2 - 0 - 1 -> 1
col: 1 -> 1 % 2 -> 1

2

3
*/